Flaky test `TestDocumentWithProjects/watch_document_with_different_projects_test`

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!
-->

**What happened**:
`TestDocumentWithProjects/watch_document_with_different_projects_test`(`test/integration/document_test.go:528`) fails intermittently. The assertion compares the watch-event sequence with `assert.Equal`, which is order-sensitive, but the `document-watched` and `document-unwatched` events are delivered asynchronously over the pub/sub stream and can arrive in either order:
```
--- FAIL: TestDocumentWithProjects/watch_document_with_different_projects_test
document_test.go:528:
Error: Not equal:
       expected: [document-watched, document-unwatched]
       actual  : [document-unwatched, document-watched]
```
Observed on main CI (run on 2026-06-12) and reproduced locally at ~12% (3/25 fresh runs) on commit 26ba66f7 (v0.7.12).

**What you expected to happen**:
The test should pass deterministically regardless of the delivery order of the watched/unwatched events, since the test only cares that both events were received with the expected presences.

**How to reproduce it (as minimally and precisely as possible)**:
With MongoDB running, run the test in fresh processes repeatedly (each run needs its own DB — see note below):
```
go test -tags integration \
    -run 'TestDocumentWithProjects/watch_document_with_different_projects_test' \
    -count=1 ./test/integration/
```
Repeat ~25 times; it fails in roughly 1 of 8 runs with the ordering mismatch above.

**Environment**:
  - Operating system: macOS (Darwin 25.0.0, arm64); also seen on Linux CI
  - Yorkie version: v0.7.12 (commit 26ba66f7)
  - Go: 1.24

--- a/client/client.go
+++ b/client/client.go
@@ -921,14 +921,65 @@ func (c *Client) runWatchLoop(ctx context.Context, d *document.Document) error {
 	rch := make(chan WatchDocResponse)
 	attachment.watchStream = rch
 
+	// The delivery pipeline has three goroutines with a single direction of
+	// backpressure:
+	//
+	//	stream reader ─┐
+	//	               ├─> watchBuffer ─> sender ─> rch
+	//	pump ──────────┘
+	//
+	// The pump is the sole consumer of the document event channel and only
+	// appends to the unbounded buffer, so producers emitting document events
+	// under the document mutex are never blocked by a slow application
+	// reading rch. The sender is the sole writer and closer of rch. The pump
+	// stops only after the stream reader has exited, so a reconcile emission
+	// in flight always has a live consumer.
+	buf := newWatchBuffer()
+	pumpStop := make(chan struct{})
+
+	// pump: document events -> buf.
+	go func() {
+		for {
+			select {
+			case e := <-d.Events():
+				t := PresenceChanged
+				if e.Type == document.WatchedEvent {
+					t = DocumentWatched
+				} else if e.Type == document.UnwatchedEvent {
+					t = DocumentUnwatched
+				}
+				buf.push(WatchDocResponse{Type: t, Presences: e.Presences})
+			case <-pumpStop:
+				return
+			}
+		}
+	}()
+
+	// sender: buf -> rch.
+	go func() {
+		defer close(rch)
+		for {
+			resp, ok := buf.pop(ctx)
+			if !ok {
+				return
+			}
+			select {
+			case rch <- resp:
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	// stream reader: server responses -> buf.
 	go func() {
 		for stream.Receive() {
 			pbResp := stream.Msg()
 			resp, err := handleWatchResponse(pbResp, d)
 			if err != nil {
-				rch <- WatchDocResponse{Err: err}
-				ctx.Done()
-				close(rch)
+				close(pumpStop)
+				buf.push(WatchDocResponse{Err: err})
+				buf.close()
 				return
 			}
 			if resp == nil {
@@ -945,46 +996,21 @@ func (c *Client) runWatchLoop(ctx context.Context, d *document.Document) error {
 				}
 			}
 
-			rch <- *resp
+			buf.push(*resp)
 		}
-		if err = stream.Err(); err != nil {
-			rch <- WatchDocResponse{Err: err}
-			ctx.Done()
-			close(rch)
 
-			// If watch stream is disconnected, we re-establish the watch stream.
-			err = c.runWatchLoop(ctx, d)
-			if err != nil {
-				return
-			}
+		close(pumpStop)
+		if err := stream.Err(); err != nil {
+			buf.push(WatchDocResponse{Err: err})
+			buf.close()
+
+			// If watch stream is disconnected, we re-establish the watch
+			// stream. The pump above has already stopped, so the new loop's
+			// pump is the sole consumer of the document event channel.
+			_ = c.runWatchLoop(ctx, d)
 			return
 		}
-	}()
-
-	// TODO(hackerwins): We need to revise the implementation of the watch
-	// event handling. Currently, we are using the same channel for both
-	// document events and watch events. This is not ideal because the
-	// client cannot distinguish between document events and watch events.
-	// We'll expose only document events and watch events will be handled
-	// internally.
-
-	// TODO(hackerwins): We should ensure that the goroutine is closed when
-	// the stream is closed.
-	go func() {
-		for {
-			select {
-			case e := <-d.Events():
-				t := PresenceChanged
-				if e.Type == document.WatchedEvent {
-					t = DocumentWatched
-				} else if e.Type == document.UnwatchedEvent {
-					t = DocumentUnwatched
-				}
-				rch <- WatchDocResponse{Type: t, Presences: e.Presences}
-			case <-ctx.Done():
-				return
-			}
-		}
+		buf.close()
 	}()
 
 	return nil
@@ -1025,27 +1051,15 @@ func handleWatchResponse(pbResp *api.WatchResponse, d *document.Document) (*Watc
 			case events.DocChanged:
 				return &WatchDocResponse{Type: DocumentChanged}, nil
 			case events.DocWatched:
-				event := d.AddOnlineClientAndReconcile(cli.String())
-				if event == nil {
-					return nil, nil
-				}
-				t := PresenceChanged
-				if event.Type == document.WatchedEvent {
-					t = DocumentWatched
-				}
-				return &WatchDocResponse{
-					Type:      t,
-					Presences: event.Presences,
-				}, nil
+				// NOTE(hackerwins): The reconciled event is emitted through
+				// the document event channel and forwarded by the event
+				// goroutine in runWatchLoop, keeping presence lifecycle
+				// events on a single ordered path (yorkie#1847).
+				d.AddOnlineClientAndReconcile(cli.String())
+				return nil, nil
 			case events.DocUnwatched:
-				event := d.RemoveOnlineClientAndReconcile(cli.String())
-				if event == nil {
-					return nil, nil
-				}
-				return &WatchDocResponse{
-					Type:      DocumentUnwatched,
-					Presences: event.Presences,
-				}, nil
+				d.RemoveOnlineClientAndReconcile(cli.String())
+				return nil, nil
 			}
 		}
 	}

--- a/client/watch_buffer.go
+++ b/client/watch_buffer.go
@@ -1,0 +1,98 @@
+/*
+ * Copyright 2026 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package client
+
+import (
+	"context"
+	"sync"
+)
+
+// watchBuffer is an unbounded FIFO queue between watch event producers and
+// the single goroutine that delivers responses to the user-facing watch
+// response channel. Push never blocks, so producers holding the document
+// mutex are never coupled to how fast the application consumes responses.
+type watchBuffer struct {
+	mu     sync.Mutex
+	items  []WatchDocResponse
+	closed bool
+
+	// signal wakes a waiting pop; capacity 1 so push never blocks.
+	signal chan struct{}
+}
+
+// newWatchBuffer creates a new watchBuffer.
+func newWatchBuffer() *watchBuffer {
+	return &watchBuffer{
+		signal: make(chan struct{}, 1),
+	}
+}
+
+// push appends the given response to the buffer. It never blocks. Pushing
+// to a closed buffer is a no-op.
+func (b *watchBuffer) push(resp WatchDocResponse) {
+	b.mu.Lock()
+	if b.closed {
+		b.mu.Unlock()
+		return
+	}
+	b.items = append(b.items, resp)
+	b.mu.Unlock()
+
+	select {
+	case b.signal <- struct{}{}:
+	default:
+	}
+}
+
+// close marks the buffer as closed. Items already buffered can still be
+// drained by pop; once drained, pop returns false.
+func (b *watchBuffer) close() {
+	b.mu.Lock()
+	b.closed = true
+	b.mu.Unlock()
+
+	select {
+	case b.signal <- struct{}{}:
+	default:
+	}
+}
+
+// pop removes and returns the oldest response. It blocks until an item is
+// available, the buffer is closed and drained, or the context is done. The
+// second return value is false when there is nothing more to deliver.
+func (b *watchBuffer) pop(ctx context.Context) (WatchDocResponse, bool) {
+	for {
+		b.mu.Lock()
+		if len(b.items) > 0 {
+			resp := b.items[0]
+			b.items = b.items[1:]
+			b.mu.Unlock()
+			return resp, true
+		}
+		if b.closed {
+			b.mu.Unlock()
+			return WatchDocResponse{}, false
+		}
+		b.mu.Unlock()
+
+		select {
+		case <-b.signal:
+		case <-ctx.Done():
+			return WatchDocResponse{}, false
+		}
+	}
+}

--- a/client/watch_buffer_test.go
+++ b/client/watch_buffer_test.go
@@ -1,0 +1,95 @@
+/*
+ * Copyright 2026 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package client
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestWatchBuffer(t *testing.T) {
+	t.Run("push never blocks and pop preserves FIFO order", func(t *testing.T) {
+		buf := newWatchBuffer()
+		types := []WatchDocResponseType{
+			DocumentWatched, DocumentChanged, DocumentUnwatched,
+		}
+		for _, tp := range types {
+			buf.push(WatchDocResponse{Type: tp})
+		}
+
+		ctx := context.Background()
+		for _, tp := range types {
+			resp, ok := buf.pop(ctx)
+			assert.True(t, ok)
+			assert.Equal(t, tp, resp.Type)
+		}
+	})
+
+	t.Run("pop blocks until push arrives", func(t *testing.T) {
+		buf := newWatchBuffer()
+		done := make(chan WatchDocResponse, 1)
+		go func() {
+			resp, ok := buf.pop(context.Background())
+			assert.True(t, ok)
+			done <- resp
+		}()
+
+		buf.push(WatchDocResponse{Type: DocumentWatched})
+		select {
+		case resp := <-done:
+			assert.Equal(t, DocumentWatched, resp.Type)
+		case <-time.After(time.Second):
+			t.Fatal("pop did not receive pushed item")
+		}
+	})
+
+	t.Run("close drains remaining items then reports exhaustion", func(t *testing.T) {
+		buf := newWatchBuffer()
+		buf.push(WatchDocResponse{Type: DocumentWatched})
+		buf.close()
+		buf.push(WatchDocResponse{Type: DocumentChanged}) // dropped: closed
+
+		ctx := context.Background()
+		resp, ok := buf.pop(ctx)
+		assert.True(t, ok)
+		assert.Equal(t, DocumentWatched, resp.Type)
+
+		_, ok = buf.pop(ctx)
+		assert.False(t, ok)
+	})
+
+	t.Run("pop returns on context cancellation", func(t *testing.T) {
+		buf := newWatchBuffer()
+		ctx, cancel := context.WithCancel(context.Background())
+		done := make(chan bool, 1)
+		go func() {
+			_, ok := buf.pop(ctx)
+			done <- ok
+		}()
+
+		cancel()
+		select {
+		case ok := <-done:
+			assert.False(t, ok)
+		case <-time.After(time.Second):
+			t.Fatal("pop did not return after context cancellation")
+		}
+	})
+}

--- a/docs/design/doc-presence.md
+++ b/docs/design/doc-presence.md
@@ -527,3 +527,121 @@ Both orderings produce the same result through the same code path.
 | Existing tests depend on current (buggy) event timing | Update tests to expect correct behavior — old timing was a bug |
 | Edge cases missed in reconcile | State space is small (2 bits × 2 bits = 16 transitions) — exhaustive verification possible |
 | Behavioral difference between JS and Go SDK during rollout | JS SDK ships first as reference implementation; Go SDK follows with same pattern |
+
+## Presence Event Delivery Unification (Go SDK)
+
+> target-version: 0.7.13
+
+### Summary
+
+The reconciliation design above unified event *decisions* — which
+event to emit for a given state transition. On the Go client, event
+*delivery* remained dual-pathed, which reorders events under
+concurrency ([yorkie#1847](https://github.com/yorkie-team/yorkie/issues/1847)).
+This section unifies delivery: `Document` becomes the single publisher
+of presence lifecycle events, and delivery order is guaranteed to
+match state transition order.
+
+### Problem
+
+Reconciled events reach the user-facing watch response channel (`rch`
+in `client.runWatchLoop`) via two independent paths:
+
+1. **Stream path**: server `DocWatched`/`DocUnwatched` →
+   `handleWatchResponse` → `AddOnlineClientAndReconcile`/
+   `RemoveOnlineClientAndReconcile` returns the event → the stream
+   goroutine sends it to `rch` directly.
+2. **Sync path**: PushPull applies remote presence changes →
+   `InternalDocument.ApplyChanges` → `ReconcilePresence` → the
+   document event channel (`d.events`) → a forwarder goroutine sends
+   it to `rch`.
+
+Which path emits `watched` depends on arrival order: if the peer's
+watch event arrives before its presence syncs, reconcile returns nil
+on the stream path (waiting state) and the `watched` event is emitted
+later by the sync path. The subsequent `unwatched` still goes via the
+stream path. Two goroutines then race on the unbuffered `rch`, and
+the Go runtime picks a sender arbitrarily — `unwatched` can be
+delivered before `watched`.
+
+### Design
+
+`Document` is the single publisher of presence lifecycle events, and
+the client decouples event production from application consumption.
+
+**Emission** — `AddOnlineClientAndReconcile` and
+`RemoveOnlineClientAndReconcile` emit the reconciled event into
+`d.events` while holding `d.mu` (the same pattern `applyChanges`
+already uses) instead of returning it to the caller.
+`handleWatchResponse` no longer builds watch responses for
+`DocWatched`/`DocUnwatched`; it only updates state via the reconcile
+methods.
+
+**Delivery** — `runWatchLoop` runs a three-goroutine pipeline with a
+single direction of backpressure:
+
+```
+stream reader ─┐
+               ├─> watchBuffer (unbounded) ─> sender ─> rch
+pump ──────────┘
+```
+
+- The *pump* is the sole consumer of `d.Events()` and only appends to
+  the unbounded `watchBuffer` — it never blocks on the application.
+- The *sender* is the sole writer and closer of `rch`.
+- The *stream reader* pushes `DocumentChanged` and error responses
+  into the same buffer and drives the lifecycle: on exit it stops the
+  pump, closes the buffer, and (on stream error) re-establishes the
+  loop — the old pump is stopped before the new one starts, so
+  `d.Events()` always has exactly one consumer.
+
+**Ordering guarantee**: sends to `d.events` happen under `d.mu`, so
+lock acquisition order = state transition order = channel send order,
+and the pump/buffer/sender chain preserves FIFO. An `unwatched` event
+requires `hadPresence && wasOnline`, a state only reachable through
+the transition that emitted `watched` — so `watched` before
+`unwatched` holds structurally.
+
+**Deadlock analysis**: a producer emitting under `d.mu` waits only
+for the pump to receive, and the pump's only other action is a
+non-blocking buffer append — it never acquires `d.mu` and never
+touches `rch`. An application goroutine that reads `rch` and calls
+into the document (`Update`, `Sync`) therefore cannot form a wait
+cycle with producers: the sender absorbs `rch` backpressure into the
+buffer instead of propagating it to the document mutex. A naive
+variant of this design — sending on the buffer-1 `d.events` under
+`d.mu` with a forwarder that writes `rch` directly — deadlocks
+exactly there (consumer in `Update` waits on `d.mu`, stream reader
+holds `d.mu` waiting on the full channel, forwarder waits on `rch`),
+which is why the unbounded buffer sits between them.
+
+**Buffer growth**: the buffer is bounded in practice by peer presence
+churn between application reads; entries are small (event type plus a
+presence map reference). The buffer lives per watch-loop invocation
+and is dropped on stream teardown.
+
+### Non-Goals
+
+- The JS SDK already delivers all presence events through the single
+  `doc.publish` path; no change needed there.
+- The broader TODO of hiding watch events behind document events
+  entirely (reworking the `WatchStream` API) is out of scope.
+- Cross-category ordering between `DocumentChanged` and presence
+  lifecycle events is not guaranteed — the server-side batch
+  publisher already reorders and dedups across these categories
+  (see `doc_subscription.go`), and applications must already handle
+  a `DocumentChanged` arriving before the corresponding `watched`.
+
+### Behavioral note: fast join/leave produces no events
+
+Per the reconciliation state table, a peer that unwatches before its
+presence has synced to the observer is never surfaced: the `watched`
+condition (`hasPresence && isOnline`) was never met, so neither
+`watched` nor `unwatched` fires. This matches the JS SDK. The old
+delivery path masked this window — the stream reader was throttled by
+the blocking response-channel handoff, so the observer's sync usually
+applied the peer's presence before the unwatch was processed. With the
+pipeline the stream reader is no longer throttled, so tests (and
+applications) that want to observe a leave must first observe the
+join; `watch document with different projects test` waits for the
+`watched` event before unwatching for exactly this reason.

--- a/docs/tasks/active/20260703-watch-event-single-path-lessons.md
+++ b/docs/tasks/active/20260703-watch-event-single-path-lessons.md
@@ -1,0 +1,31 @@
+# Lessons: watch event single path
+
+**Created**: 2026-07-03
+
+(Learnings captured during the task; finalized before archive.)
+
+- The `ReconcilePresence` state machine (0.7.6) unified event
+  *decisions* but left event *delivery* dual-pathed on the Go client —
+  a reconciliation design needs to cover both to be race-free.
+- "Same pattern as applyChanges" was a false safety argument for
+  sending on the buffer-1 event channel under `d.mu`: applyChanges is
+  driven by the app's own Sync call, while reconcile emission is
+  driven externally by the watch stream. An external producer blocking
+  under the document mutex forms a wait cycle with any consumer that
+  reenters the document (`Update`/`Sync`) while handling events.
+  Deadlock analysis must trace the full consumer chain, not just the
+  producer side.
+- Delivery pipelines need explicit ownership rules: exactly one
+  consumer per channel (`d.Events()` → pump), exactly one
+  writer/closer per channel (sender → `rch`), and teardown ordered so
+  producers always outlive their consumer (stream reader exits before
+  pump stops). The pre-existing reconnect path violated all three.
+- Backpressure from a user-facing channel must be absorbed by an
+  unbounded buffer before it reaches any lock-holding producer.
+- Removing backpressure changes timing that tests silently rely on:
+  the unthrottled stream reader processed unwatch before the
+  observer's sync applied the peer's presence, so the reconcile state
+  machine (correctly, JS-parity) emitted nothing and the test hung on
+  its wait group. The goroutine dump made this obvious — always grab
+  the dump instead of guessing at a hang. Tests asserting join/leave
+  pairs must observe the join before triggering the leave.

--- a/docs/tasks/active/20260703-watch-event-single-path-todo.md
+++ b/docs/tasks/active/20260703-watch-event-single-path-todo.md
@@ -1,0 +1,64 @@
+# Unify client watch event delivery into a single path
+
+**Created**: 2026-07-03
+
+## Goal
+
+Fix the root cause of flaky
+`TestDocumentWithProjects/watch_document_with_different_projects_test`
+(#1847): `DocumentWatched`/`DocumentUnwatched` can be delivered out of
+order because two goroutines race on the same watch response channel.
+
+## Root Cause
+
+Presence lifecycle events reach the user-facing `rch` channel via two
+independent paths in `client.runWatchLoop`:
+
+1. Stream path: server `DocWatched`/`DocUnwatched` → `handleWatchResponse`
+   → reconcile → stream goroutine sends directly to `rch`.
+2. Sync path: PushPull applies presence changes → `ApplyChanges` →
+   `ReconcilePresence` → `d.events` → forwarder goroutine sends to `rch`.
+
+When c2's watch event arrives before its presence syncs, the
+`WatchedEvent` is emitted later via path 2 while the `UnwatchedEvent`
+goes via path 1 — two goroutines race on unbuffered `rch`, reordering
+events ~12% of runs.
+
+## Design
+
+`Document` becomes the single publisher of presence lifecycle events.
+See `docs/design/doc-presence.md` — "Presence Event Delivery
+Unification" section.
+
+- `AddOnlineClientAndReconcile`/`RemoveOnlineClientAndReconcile` emit
+  the reconciled event into `d.events` while holding `d.mu` (same
+  pattern as `applyChanges`), instead of returning it. Lock order =
+  state transition order = channel send order.
+- `handleWatchResponse` returns `nil, nil` for `DocWatched`/
+  `DocUnwatched`; the `d.Events()` forwarder becomes the sole source
+  of watched/unwatched/presence-changed on `rch`.
+
+## Checklist
+
+- [x] Update `docs/design/doc-presence.md` with delivery unification
+      section
+- [x] `pkg/document/document.go`: reconcile methods emit into
+      `d.events` under lock, drop return value
+- [x] `client/client.go`: `handleWatchResponse` stops sending
+      watched/unwatched directly
+- [x] Keep `document_test.go` order-sensitive assert as regression
+      guard
+- [x] Self-review round 1: found send-under-lock deadlock class
+      (producer under `d.mu` → cap-1 `d.events` → forwarder → `rch` →
+      consumer calling `Update`), stranded reconcile send on teardown,
+      and dual-forwarder panic on reconnect
+- [x] Address review: unbounded `watchBuffer` + pump/sender/stream
+      reader pipeline in `runWatchLoop`; pump stops only after stream
+      reader exits; sender is sole writer/closer of `rch`
+- [x] Fix test timing assumption surfaced by the pipeline: the
+      unthrottled stream reader can process unwatch before presence
+      syncs → no events by design (JS parity); the test now waits for
+      `watched` before unwatching
+- [ ] Verify: 20+ repeated runs of flaky + regression tests,
+      `make lint`, `go test ./...`, full integration suite
+- [ ] Open PR referencing #1847

--- a/pkg/document/document.go
+++ b/pkg/document/document.go
@@ -526,8 +526,14 @@ func (d *Document) RemoveOnlineClient(clientID string) {
 }
 
 // AddOnlineClientAndReconcile adds the given client to the online clients
-// and returns the appropriate presence event based on the state transition.
-func (d *Document) AddOnlineClientAndReconcile(clientID string) *DocEvent {
+// and emits the appropriate presence event based on the state transition
+// through the document event channel.
+//
+// The send happens while holding d.mu — the same pattern as applyChanges —
+// so that lock acquisition order, state transition order, and channel send
+// order coincide. Delivering through the caller instead reintroduces the
+// reordering race between the watch-stream and sync paths (yorkie#1847).
+func (d *Document) AddOnlineClientAndReconcile(clientID string) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
@@ -537,13 +543,16 @@ func (d *Document) AddOnlineClientAndReconcile(clientID string) *DocEvent {
 
 	d.doc.AddOnlineClient(clientID)
 
-	return d.doc.ReconcilePresence(clientID, hadPresence, wasOnline, prevPresence)
+	if event := d.doc.ReconcilePresence(clientID, hadPresence, wasOnline, prevPresence); event != nil {
+		d.events <- *event
+	}
 }
 
 // RemoveOnlineClientAndReconcile removes the given client from the online
-// clients and returns the appropriate presence event based on the state
-// transition.
-func (d *Document) RemoveOnlineClientAndReconcile(clientID string) *DocEvent {
+// clients and emits the appropriate presence event based on the state
+// transition through the document event channel. See
+// AddOnlineClientAndReconcile for why the send happens under the lock.
+func (d *Document) RemoveOnlineClientAndReconcile(clientID string) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 
@@ -553,7 +562,9 @@ func (d *Document) RemoveOnlineClientAndReconcile(clientID string) *DocEvent {
 
 	d.doc.RemoveOnlineClient(clientID)
 
-	return d.doc.ReconcilePresence(clientID, hadPresence, wasOnline, prevPresence)
+	if event := d.doc.ReconcilePresence(clientID, hadPresence, wasOnline, prevPresence); event != nil {
+		d.events <- *event
+	}
 }
 
 // Events returns the events of this document.

--- a/pkg/document/presence_event_test.go
+++ b/pkg/document/presence_event_test.go
@@ -1,0 +1,135 @@
+/*
+ * Copyright 2026 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package document_test
+
+import (
+	"testing"
+	gotime "time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/yorkie-team/yorkie/pkg/document"
+	"github.com/yorkie-team/yorkie/pkg/document/change"
+	"github.com/yorkie-team/yorkie/pkg/document/json"
+	"github.com/yorkie-team/yorkie/pkg/document/presence"
+	"github.com/yorkie-team/yorkie/pkg/document/time"
+)
+
+// receiveEvent reads one event from the document event channel or fails
+// the test after a timeout, so a missing emission does not hang the run.
+func receiveEvent(t *testing.T, doc *document.Document) document.DocEvent {
+	t.Helper()
+	select {
+	case e := <-doc.Events():
+		return e
+	case <-gotime.After(gotime.Second):
+		t.Fatal("expected an event on doc.Events(), got none")
+		return document.DocEvent{}
+	}
+}
+
+// assertNoEvent asserts that no event is pending on the document event
+// channel.
+func assertNoEvent(t *testing.T, doc *document.Document) {
+	t.Helper()
+	select {
+	case e := <-doc.Events():
+		t.Fatalf("expected no event, got %s", e.Type)
+	default:
+	}
+}
+
+// presencePack builds a change pack carrying the peer's initial presence,
+// with the version vector adjusted for the receiver as in real delivery.
+func presencePack(t *testing.T, from, to *document.Document) *change.Pack {
+	t.Helper()
+	pack := from.CreateChangePack()
+	pCopy := *pack
+	pCopy.VersionVector = pack.VersionVector.DeepCopy()
+	pCopy.VersionVector.Set(to.ActorID(),
+		to.VersionVector().VersionOf(to.ActorID()))
+	return &pCopy
+}
+
+// TestPresenceEventSinglePath pins the contract from
+// docs/design/doc-presence.md ("Presence Event Delivery Unification"):
+// AddOnlineClientAndReconcile and RemoveOnlineClientAndReconcile emit
+// their reconciled events through the document event channel, so watched
+// and unwatched are delivered in state-transition order regardless of
+// whether the peer's presence or watch signal arrives first. Previously
+// the watch-stream path returned events to the caller, which raced with
+// the sync path on the client watch response channel (yorkie#1847).
+func TestPresenceEventSinglePath(t *testing.T) {
+	newPeers := func(t *testing.T) (observer, peer *document.Document) {
+		observerActor, err := time.ActorIDFromHex("000000000000000000000001")
+		assert.NoError(t, err)
+		peerActor, err := time.ActorIDFromHex("000000000000000000000002")
+		assert.NoError(t, err)
+
+		observer = document.New("single-path")
+		observer.SetActor(observerActor)
+
+		peer = document.New("single-path")
+		peer.SetActor(peerActor)
+		assert.NoError(t, peer.Update(func(root *json.Object, p *presence.Presence) error {
+			p.Set("name", "peer")
+			return nil
+		}))
+		return observer, peer
+	}
+
+	t.Run("presence before watch", func(t *testing.T) {
+		observer, peer := newPeers(t)
+
+		// 01. Peer's presence syncs first: both parts are not ready yet,
+		// so no event is emitted (waiting for online).
+		assert.NoError(t, observer.ApplyChangePack(presencePack(t, peer, observer)))
+		assertNoEvent(t, observer)
+
+		// 02. The watch signal arrives: the reconciled watched event must
+		// come through the document event channel.
+		observer.AddOnlineClientAndReconcile(peer.ActorID().String())
+		e := receiveEvent(t, observer)
+		assert.Equal(t, document.WatchedEvent, e.Type)
+
+		// 03. The unwatch signal arrives: same channel, after watched.
+		observer.RemoveOnlineClientAndReconcile(peer.ActorID().String())
+		e = receiveEvent(t, observer)
+		assert.Equal(t, document.UnwatchedEvent, e.Type)
+	})
+
+	t.Run("watch before presence", func(t *testing.T) {
+		observer, peer := newPeers(t)
+
+		// 01. The watch signal arrives first: no presence data yet, so no
+		// event is emitted (waiting for presence).
+		observer.AddOnlineClientAndReconcile(peer.ActorID().String())
+		assertNoEvent(t, observer)
+
+		// 02. Peer's presence syncs: the watched event is emitted by the
+		// apply path through the same channel.
+		assert.NoError(t, observer.ApplyChangePack(presencePack(t, peer, observer)))
+		e := receiveEvent(t, observer)
+		assert.Equal(t, document.WatchedEvent, e.Type)
+
+		// 03. The unwatch signal arrives: unwatched must follow watched on
+		// the same channel — this ordering was racy before unification.
+		observer.RemoveOnlineClientAndReconcile(peer.ActorID().String())
+		e = receiveEvent(t, observer)
+		assert.Equal(t, document.UnwatchedEvent, e.Type)
+	})
+}

--- a/test/integration/document_test.go
+++ b/test/integration/document_test.go
@@ -452,6 +452,12 @@ func TestDocumentWithProjects(t *testing.T) {
 		wg := sync.WaitGroup{}
 		wg.Add(1)
 
+		// watchedReceived signals that c1 observed c2 joining. The unwatch
+		// below must happen after this: a peer that leaves before its
+		// presence has synced is never surfaced (see ReconcilePresence),
+		// so unwatching earlier would legitimately produce no events.
+		watchedReceived := make(chan struct{})
+
 		d1 := document.New(helper.TestKey(t))
 		err = c1.Attach(ctx, d1, client.WithRealtimeSync())
 		assert.NoError(t, err)
@@ -477,6 +483,9 @@ func TestDocumentWithProjects(t *testing.T) {
 						Type:      resp.Type,
 						Presences: resp.Presences,
 					})
+					if len(responsePairs) == 1 {
+						close(watchedReceived)
+					}
 				}
 				if len(responsePairs) == 2 {
 					return
@@ -513,6 +522,10 @@ func TestDocumentWithProjects(t *testing.T) {
 			return nil
 		}))
 		assert.NoError(t, c3.Sync(ctx))
+
+		// Wait until c1 observed c2 joining before unwatching; leaving
+		// before the join is observed produces no events by design.
+		<-watchedReceived
 
 		// c2 unwatch the document, so c1 receives a document unwatched event.
 		expected = append(expected, watchDocPair{

--- a/test/integration/watch_pipeline_test.go
+++ b/test/integration/watch_pipeline_test.go
@@ -1,0 +1,100 @@
+//go:build integration
+
+/*
+ * Copyright 2026 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package integration
+
+import (
+	"context"
+	"testing"
+	gotime "time"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/yorkie-team/yorkie/client"
+	"github.com/yorkie-team/yorkie/pkg/document"
+	"github.com/yorkie-team/yorkie/pkg/document/json"
+	"github.com/yorkie-team/yorkie/pkg/document/presence"
+	"github.com/yorkie-team/yorkie/test/helper"
+)
+
+// TestWatchEventDeliveryPipeline guards the delivery pipeline described in
+// docs/design/doc-presence.md ("Presence Event Delivery Unification").
+func TestWatchEventDeliveryPipeline(t *testing.T) {
+	clients := activeClients(t, 2)
+	c1, c2 := clients[0], clients[1]
+	defer deactivateAndCloseClients(t, clients)
+
+	// A watch consumer that re-enters the document (Update acquires the
+	// document mutex) on every event must not deadlock against presence
+	// event producers. Before the pipeline, the stream goroutine emitted
+	// into a full buffer-1 channel while holding the document mutex, the
+	// forwarder waited on the response channel, and a consumer inside
+	// Update waited on the mutex — a three-way cycle under peer churn.
+	t.Run("consumer reentering document under peer churn test", func(t *testing.T) {
+		ctx := context.Background()
+
+		d1 := document.New(helper.TestKey(t))
+		assert.NoError(t, c1.Attach(ctx, d1, client.WithRealtimeSync()))
+		rch, cancel1, err := c1.WatchStream(d1)
+		assert.NoError(t, err)
+
+		consumerDone := make(chan struct{})
+		go func() {
+			defer close(consumerDone)
+			for resp := range rch {
+				if resp.Err != nil {
+					return
+				}
+				// Reenter the document on every event to hold the document
+				// mutex while further events are being produced.
+				assert.NoError(t, d1.Update(func(root *json.Object, p *presence.Presence) error {
+					p.Set("seen", "yes")
+					return nil
+				}))
+				if resp.Type == client.DocumentChanged {
+					assert.NoError(t, c1.Sync(ctx, client.WithKey(d1.Key())))
+				}
+			}
+		}()
+
+		// Peer churn: repeated join, presence update, and leave produce
+		// bursts of watched/unwatched/presence-changed events.
+		for i := 0; i < 5; i++ {
+			d2 := document.New(d1.Key())
+			assert.NoError(t, c2.Attach(ctx, d2, client.WithRealtimeSync()))
+			_, cancel2, err := c2.WatchStream(d2)
+			assert.NoError(t, err)
+
+			assert.NoError(t, d2.Update(func(root *json.Object, p *presence.Presence) error {
+				p.Set("round", "r")
+				return nil
+			}))
+			assert.NoError(t, c2.Sync(ctx))
+
+			cancel2()
+			assert.NoError(t, c2.Detach(ctx, d2))
+		}
+
+		cancel1()
+		select {
+		case <-consumerDone:
+		case <-gotime.After(30 * gotime.Second):
+			t.Fatal("watch consumer deadlocked against presence event producers")
+		}
+	})
+}

